# Indiana University Health

https://iuhealth.org/schedule-an-appointment/price-estimates/standard-charge-listing 

Standard Charge Listing for IU Health Methodist, IU Health Morgan, IU Health Saxony, IU Health University, and Riley Hospital for Children at IU Health:

https://cdn.iuhealth.org/resources/351955872_indiana-university-health-inc._standardcharges.zip 


In [2]:
# Load duckdb, which lets us efficiently load large files
import duckdb

# Load pandas, which lets us manipulate dataframes
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

# Set configrations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.
%config SqlMagic.autopandas = True

%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

# Allow named parameters (python variables) in SQL cells
%config SqlMagic.named_parameters=True

# Connect jupysql to DuckDB using a SQLAlchemy-style connection string. Either connect to an in memory DuckDB, or a file backed db.
%sql duckdb:///:memory:

In [3]:
!wget https://cdn.iuhealth.org/resources/351955872_indiana-university-health-inc._standardcharges.zip -P /tmp

--2023-09-07 14:48:36--  https://cdn.iuhealth.org/resources/351955872_indiana-university-health-inc._standardcharges.zip
Resolving cdn.iuhealth.org (cdn.iuhealth.org)... 3.162.174.57, 3.162.174.17, 3.162.174.36, ...
Connecting to cdn.iuhealth.org (cdn.iuhealth.org)|3.162.174.57|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 69872858 (67M) [application/zip]
Saving to: ‘/tmp/351955872_indiana-university-health-inc._standardcharges.zip’

351955872_indiana-u 100%[===================>]  66.64M  9.74MB/s    in 8.0s    

2023-09-07 14:48:44 (8.34 MB/s) - ‘/tmp/351955872_indiana-university-health-inc._standardcharges.zip’ saved [69872858/69872858]



In [6]:
!unzip /tmp/351955872_indiana-university-health-inc._standardcharges.zip -d ~/data/payless_health

Archive:  /tmp/351955872_indiana-university-health-inc._standardcharges.zip
  inflating: /Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv  


In [7]:
!head /Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv

Service Code,Revenue Code,NDC,Modifier,CPT/HCPCS/DRG,Code Type,Description,Patient Type,Contract,Gross Charge,Payer-specific negotiated rate,Rate Type,De-identified maximum negotiated charge,De-identified minimum negotiated charge,Discounted cash price
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Medicaid_Managed Health Services_HIP",64,37.21,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","SIHO_Open Access PPO",64,53.76,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","SIHO_Inspire",64,41.6,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Parkview Signature Care PPO",64,51.2,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Community 

# Prompt for Claude for loading data

````
```
SELECT *
FROM read_csv('https://data.cityofnewyork.us/api/views/erm2-nwe9/rows.csv?accessType=DOWNLOAD',
    header=True,
    delim=',',
    quote='"',
    columns={'Unique Key': 'BIGINT',
    'Created Date': 'VARCHAR',
    'Closed Date': 'VARCHAR',
    'Agency': 'VARCHAR',
    'Agency Name': 'VARCHAR',
    'Complaint Type': 'VARCHAR',
    'Descriptor': 'VARCHAR',
    'Location Type': 'VARCHAR',
    'Incident Zip': 'VARCHAR',
    'Incident Address': 'VARCHAR',
    'Street Name': 'VARCHAR',
    'Cross Street 1': 'VARCHAR',
    'Cross Street 2': 'VARCHAR',
    'Intersection Street 1': 'VARCHAR',
    'Intersection Street 2': 'VARCHAR',
    'Address Type': 'VARCHAR',
    'City': 'VARCHAR',
    'Landmark': 'VARCHAR',
    'Facility Type': 'VARCHAR',
    'Status': 'VARCHAR',
    'Due Date': 'VARCHAR',
    'Resolution Description': 'VARCHAR',
    'Resolution Action Updated Date': 'VARCHAR',
    'Community Board': 'VARCHAR',
    'BBL': 'VARCHAR',
    'Borough': 'VARCHAR',
    'X Coordinate (State Plane)': 'VARCHAR',
    'Y Coordinate (State Plane)': 'VARCHAR',
    'Open Data Channel Type': 'VARCHAR',
    'Park Facility Name': 'VARCHAR',
    'Park Borough': 'VARCHAR',
    'Vehicle Type': 'VARCHAR',
    'Taxi Company Borough': 'VARCHAR',
    'Taxi Pick Up Location': 'VARCHAR',
    'Bridge Highway Name': 'VARCHAR',
    'Bridge Highway Direction': 'VARCHAR',
    'Road Ramp': 'VARCHAR',
    'Bridge Highway Segment': 'VARCHAR',
    'Latitude': 'DOUBLE',
    'Longitude': 'DOUBLE',
    'Location': 'VARCHAR'}) 
LIMIT 10;
```

Please use the above query and rewrite it for the file `/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv` with the following header:

```
Service Code,Revenue Code,NDC,Modifier,CPT/HCPCS/DRG,Code Type,Description,Patient Type,Contract,Gross Charge,Payer-specific negotiated rate,Rate Type,De-identified maximum negotiated charge,De-identified minimum negotiated charge,Discounted cash price
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Medicaid_Managed Health Services_HIP",64,37.21,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","SIHO_Open Access PPO",64,53.76,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","SIHO_Inspire",64,41.6,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Parkview Signature Care PPO",64,51.2,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Community Health Direct Other",64,44.8,"Dollar Amount",107.52,20.14,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","AETNA_New Business",64,33.86,"Dollar Amount",107.52,20.14,20.14
"57720070","771","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","SAGAMORE_NON DRG",64,47.62,"Dollar Amount",107.52,0,20.14
"57720070","636","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Multiplan",64,57.6,"Dollar Amount",107.52,20.14,20.14
"57720070","771","-1","-1","0001A","CPT/HCPCS","ADM SARSCOV2 30MCG 0.3ML 1ST","Outpatient","Parkview Signature Care PPO",64,51.2,"Dollar Amount",107.52,0,20.14
```

````


In [8]:
%%sql 
SELECT *
FROM read_csv('/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv',
    header=True,
    delim=',',
    quote='"',
    columns={'Service Code': 'VARCHAR',
             'Revenue Code': 'VARCHAR',
             'NDC': 'VARCHAR',
             'Modifier': 'VARCHAR',
             'CPT/HCPCS/DRG': 'VARCHAR',
             'Code Type': 'VARCHAR',
             'Description': 'VARCHAR',
             'Patient Type': 'VARCHAR',
             'Contract': 'VARCHAR',
             'Gross Charge': 'DOUBLE',
             'Payer-specific negotiated rate': 'DOUBLE',
             'Rate Type': 'VARCHAR',
             'De-identified maximum negotiated charge': 'DOUBLE',
             'De-identified minimum negotiated charge': 'DOUBLE',
             'Discounted cash price': 'DOUBLE'})
LIMIT 10;

Service Code Revenue Code NDC Modifier CPT/HCPCS/DRG  Code Type  \
0     57720070          636  -1       -1         0001A  CPT/HCPCS   
1     57720070          636  -1       -1         0001A  CPT/HCPCS   
2     57720070          636  -1       -1         0001A  CPT/HCPCS   
3     57720070          636  -1       -1         0001A  CPT/HCPCS   
4     57720070          636  -1       -1         0001A  CPT/HCPCS   
5     57720070          636  -1       -1         0001A  CPT/HCPCS   
6     57720070          771  -1       -1         0001A  CPT/HCPCS   
7     57720070          636  -1       -1         0001A  CPT/HCPCS   
8     57720070          771  -1       -1         0001A  CPT/HCPCS   
9     57720070          771  -1       -1         0001A  CPT/HCPCS   

                    Description Patient Type  \
0  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
1  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
2  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
3  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
4  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
5  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
6  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
7  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
8  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
9  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   

                               Contract  Gross Charge  \
0  Medicaid_Managed Health Services_HIP          64.0   
1                  SIHO_Open Access PPO          64.0   
2                          SIHO_Inspire          64.0   
3           Parkview Signature Care PPO          64.0   
4         Community Health Direct Other          64.0   
5                    AETNA_New Business          64.0   
6                      SAGAMORE_NON DRG          64.0   
7                             Multiplan          64.0   
8           Parkview Signature Care PPO          64.0   
9                             AETNA_COX          64.0   

   Payer-specific negotiated rate      Rate Type  \
0                           37.21  Dollar Amount   
1                           53.76  Dollar Amount   
2                           41.60  Dollar Amount   
3                           51.20  Dollar Amount   
4                           44.80  Dollar Amount   
5                           33.86  Dollar Amount   
6                           47.62  Dollar Amount   
7                           57.60  Dollar Amount   
8                           51.20  Dollar Amount   
9                           33.86  Dollar Amount   

   De-identified maximum negotiated charge  \
0                                   107.52   
1                                   107.52   
2                                   107.52   
3                                   107.52   
4                                   107.52   
5                                   107.52   
6                                   107.52   
7                                   107.52   
8                                   107.52   
9                                   107.52   

   De-identified minimum negotiated charge  Discounted cash price  
0                                    20.14                  20.14  
1                                    20.14                  20.14  
2                                    20.14                  20.14  
3                                    20.14                  20.14  
4                                    20.14                  20.14  
5                                    20.14                  20.14  
6                                     0.00                  20.14  
7                                    20.14                  20.14  
8                                     0.00                  20.14  
9                                     0.00                  20.14

In [3]:
def lower_case_and_underscore(dictionary):
    return {k.lower().replace(' ', '_').replace('-', '_').replace('/', '_'): v for k, v in dictionary.items()}


In [4]:
column_names = {'Service Code': 'VARCHAR',
             'Revenue Code': 'VARCHAR',
             'NDC': 'VARCHAR',
             'Modifier': 'VARCHAR',
             'CPT/HCPCS/DRG': 'VARCHAR',
             'Code Type': 'VARCHAR',
             'Description': 'VARCHAR',
             'Patient Type': 'VARCHAR',
             'Contract': 'VARCHAR',
             'Gross Charge': 'DOUBLE',
             'Payer-specific negotiated rate': 'DOUBLE',
             'Rate Type': 'VARCHAR',
             'De-identified maximum negotiated charge': 'DOUBLE',
             'De-identified minimum negotiated charge': 'DOUBLE',
             'Discounted cash price': 'DOUBLE'}

In [5]:
lower_case_and_underscore(column_names)

{'service_code': 'VARCHAR',
 'revenue_code': 'VARCHAR',
 'ndc': 'VARCHAR',
 'modifier': 'VARCHAR',
 'cpt_hcpcs_drg': 'VARCHAR',
 'code_type': 'VARCHAR',
 'description': 'VARCHAR',
 'patient_type': 'VARCHAR',
 'contract': 'VARCHAR',
 'gross_charge': 'DOUBLE',
 'payer_specific_negotiated_rate': 'DOUBLE',
 'rate_type': 'VARCHAR',
 'de_identified_maximum_negotiated_charge': 'DOUBLE',
 'de_identified_minimum_negotiated_charge': 'DOUBLE',
 'discounted_cash_price': 'DOUBLE'}

In [6]:
%%sql 
SELECT *
FROM read_csv('/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv',
    header=True,
    delim=',',
    quote='"',
    nullstr='-1',
    columns={'service_code': 'VARCHAR',
            'revenue_code': 'VARCHAR',
            'ndc': 'VARCHAR',
            'modifier': 'VARCHAR',
            'cpt_hcpcs_drg': 'VARCHAR',
            'code_type': 'VARCHAR',
            'description': 'VARCHAR',
            'patient_type': 'VARCHAR',
            'contract': 'VARCHAR',
            'gross_charge': 'DOUBLE',
            'payer_specific_negotiated_rate': 'DOUBLE',
            'rate_type': 'VARCHAR',
            'de_identified_maximum_negotiated_charge': 'DOUBLE',
            'de_identified_minimum_negotiated_charge': 'DOUBLE',
            'discounted_cash_price': 'DOUBLE'})
LIMIT 10;

service_code revenue_code  ndc modifier cpt_hcpcs_drg  code_type  \
0     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
1     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
2     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
3     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
4     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
5     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
6     57720070          771  NaN      NaN         0001A  CPT/HCPCS   
7     57720070          636  NaN      NaN         0001A  CPT/HCPCS   
8     57720070          771  NaN      NaN         0001A  CPT/HCPCS   
9     57720070          771  NaN      NaN         0001A  CPT/HCPCS   

                    description patient_type  \
0  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
1  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
2  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
3  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
4  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
5  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
6  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
7  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
8  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
9  ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   

                               contract  gross_charge  \
0  Medicaid_Managed Health Services_HIP          64.0   
1                  SIHO_Open Access PPO          64.0   
2                          SIHO_Inspire          64.0   
3           Parkview Signature Care PPO          64.0   
4         Community Health Direct Other          64.0   
5                    AETNA_New Business          64.0   
6                      SAGAMORE_NON DRG          64.0   
7                             Multiplan          64.0   
8           Parkview Signature Care PPO          64.0   
9                             AETNA_COX          64.0   

   payer_specific_negotiated_rate      rate_type  \
0                           37.21  Dollar Amount   
1                           53.76  Dollar Amount   
2                           41.60  Dollar Amount   
3                           51.20  Dollar Amount   
4                           44.80  Dollar Amount   
5                           33.86  Dollar Amount   
6                           47.62  Dollar Amount   
7                           57.60  Dollar Amount   
8                           51.20  Dollar Amount   
9                           33.86  Dollar Amount   

   de_identified_maximum_negotiated_charge  \
0                                   107.52   
1                                   107.52   
2                                   107.52   
3                                   107.52   
4                                   107.52   
5                                   107.52   
6                                   107.52   
7                                   107.52   
8                                   107.52   
9                                   107.52   

   de_identified_minimum_negotiated_charge  discounted_cash_price  
0                                    20.14                  20.14  
1                                    20.14                  20.14  
2                                    20.14                  20.14  
3                                    20.14                  20.14  
4                                    20.14                  20.14  
5                                    20.14                  20.14  
6                                     0.00                  20.14  
7                                    20.14                  20.14  
8                                     0.00                  20.14  
9                                     0.00                  20.14

In [7]:
%%sql
COPY (
SELECT *
FROM read_csv('/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.csv',
    header=True,
    delim=',',
    quote='"',
    nullstr='-1',
    columns={'service_code': 'VARCHAR',
            'revenue_code': 'VARCHAR',
            'ndc': 'VARCHAR',
            'modifier': 'VARCHAR',
            'cpt_hcpcs_drg': 'VARCHAR',
            'code_type': 'VARCHAR',
            'description': 'VARCHAR',
            'patient_type': 'VARCHAR',
            'insurance_name': 'VARCHAR',
            'gross_charge': 'DOUBLE',
            'negotiated_rate': 'DOUBLE',
            'rate_type': 'VARCHAR',
            'de_identified_maximum_negotiated_charge': 'DOUBLE',
            'de_identified_minimum_negotiated_charge': 'DOUBLE',
            'discounted_cash_price': 'DOUBLE'})
) TO '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet' (COMPRESSION ZSTD);

Empty DataFrame
Columns: [Success]
Index: []

In [8]:
%%sql 
SELECT * FROM '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet'

service_code revenue_code  ndc modifier cpt_hcpcs_drg  code_type  \
0           57720070          636  NaN      NaN         0001A  CPT/HCPCS   
1           57720070          636  NaN      NaN         0001A  CPT/HCPCS   
2           57720070          636  NaN      NaN         0001A  CPT/HCPCS   
3           57720070          636  NaN      NaN         0001A  CPT/HCPCS   
4           57720070          636  NaN      NaN         0001A  CPT/HCPCS   
...              ...          ...  ...      ...           ...        ...   
6599487     54701990          470  NaN      NaN         V5299  CPT/HCPCS   
6599488     54701990          470  NaN       GY         V5299  CPT/HCPCS   
6599489     54701990          470  NaN       GY         V5299  CPT/HCPCS   
6599490     54701990          470  NaN       GY         V5299  CPT/HCPCS   
6599491     54701990          470  NaN       GY         V5299  CPT/HCPCS   

                          description patient_type  \
0        ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
1        ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
2        ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
3        ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
4        ADM SARSCOV2 30MCG 0.3ML 1ST   Outpatient   
...                               ...          ...   
6599487         MISC HEARING SERVICES   Outpatient   
6599488         MISC HEARING SERVICES   Outpatient   
6599489         MISC HEARING SERVICES   Outpatient   
6599490         MISC HEARING SERVICES   Outpatient   
6599491         MISC HEARING SERVICES   Outpatient   

                               insurance_name  gross_charge  negotiated_rate  \
0        Medicaid_Managed Health Services_HIP          64.0            37.21   
1                        SIHO_Open Access PPO          64.0            53.76   
2                                SIHO_Inspire          64.0            41.60   
3                 Parkview Signature Care PPO          64.0            51.20   
4               Community Health Direct Other          64.0            44.80   
...                                       ...           ...              ...   
6599487                            ENCORE_PPO          81.0            58.32   
6599488    CARESOURCE MARKETPLACE_MARKETPLACE          81.0             0.00   
6599489                         AETNA_PPO/POS          81.0            54.43   
6599490                           ANTHEM_TRAD          81.0            57.53   
6599491          Medicare_Humana_PPO_Medicare          81.0             0.00   

             rate_type  de_identified_maximum_negotiated_charge  \
0        Dollar Amount                                   107.52   
1        Dollar Amount                                   107.52   
2        Dollar Amount                                   107.52   
3        Dollar Amount                                   107.52   
4        Dollar Amount                                   107.52   
...                ...                                      ...   
6599487  Dollar Amount                                   186.46   
6599488  Dollar Amount                                   186.46   
6599489  Dollar Amount                                   186.46   
6599490  Dollar Amount                                   186.46   
6599491  Dollar Amount                                   186.46   

         de_identified_minimum_negotiated_charge  discounted_cash_price  
0                                          20.14                  20.14  
1                                          20.14                  20.14  
2                                          20.14                  20.14  
3                                          20.14                  20.14  
4                                          20.14                  20.14  
...                                          ...                    ...  
6599487                                     0.00                  25.49  
6599488                                     0.00                  25.49  
6599489       

# Prompt for Claude for creating dashboard

````
```
  WITH max_min_charges AS (
    SELECT
      billing_code,
      description,
      MIN(charge) AS min_charge,
      MAX(charge) AS max_charge
    FROM '/Users/me/data/payless_health/231689692_Lehigh_Valley_Hospital_StandardCharges.parquet'
    GROUP BY billing_code, description
  ),

  min_max_names AS (
    SELECT
      mmc.billing_code,
      mmc.description,
      mmc.min_charge,
      mmc.max_charge,
      FIRST_VALUE(payor_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY charge ASC) AS name_minimum,
      FIRST_VALUE(payor_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY charge DESC) AS name_maximum
    FROM max_min_charges mmc
    JOIN '/Users/me/data/payless_health/231689692_Lehigh_Valley_Hospital_StandardCharges.parquet' sc
    ON mmc.billing_code = sc.billing_code AND mmc.description = sc.description
  )

  SELECT DISTINCT
    billing_code,
    description,
    min_charge,
    max_charge,
    name_minimum,
    name_maximum
  FROM min_max_names
  ```

  Please use the above query and rewrite it for the file `/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet` the following column names `insurance_name, negotiated_rate, description, cpt_hcpcs_drg`
  ````
  ``

  

In [9]:
%%sql 
WITH max_min_charges AS (
  SELECT
    cpt_hcpcs_drg AS billing_code, 
    description,
    MIN(negotiated_rate) AS min_charge,
    MAX(negotiated_rate) AS max_charge
  FROM '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet'
  WHERE negotiated_rate > 0
  GROUP BY cpt_hcpcs_drg, description
),

min_max_names AS (
  SELECT
    mmc.billing_code,
    mmc.description, 
    mmc.min_charge,
    mmc.max_charge,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate ASC) AS name_minimum,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate DESC) AS name_maximum
  FROM max_min_charges mmc
  JOIN '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet' sc
  ON mmc.billing_code = sc.cpt_hcpcs_drg AND mmc.description = sc.description
)

SELECT DISTINCT
  billing_code,
  description,
  min_charge,
  max_charge,
  name_minimum,
  name_maximum
FROM min_max_names

billing_code                    description  min_charge  max_charge  \
0            0003A   ADM SARSCOV2 30MCG 0.3ML 3RD       26.24      107.52   
1            0012A  ADM SARSCOV2 100MCG 0.5ML 2ND       26.24      107.52   
2            00145       CR VITREORETINAL SURG P3       43.05       94.50   
3            0047U     Onc prst8 mrna 17 gene alg     3873.00     3873.00   
4            0061U   Tc meas 5 bmrk sfdi m-s alys       25.00       25.10   
...            ...                            ...         ...         ...   
21714        86790                       HHV 7 AB       12.88       58.50   
21715        93325   DOPPLR ECHO COLORFLOW MAP IP      387.86      851.40   
21716        C1713      ANCHOR SCREW BN/BN TIS/BN       36.08    41197.50   
21717        J1930  LANREOTIDE DEPOT INJ PER 1 MG       35.33      233.37   
21718          947    SIGNS AND SYMPTOMS WITH MCC        0.41    50519.12   

                               name_minimum               name_maximum  
0      Medicaid_Managed Health Services_HIP      IUHPS Narrow Network_  
1           Medicaid_Miscellaneous_Medicaid      IUHPS Narrow Network_  
2                       Medicaid_Anthem_HIP  Community Health Alliance  
3      Medicaid_Managed Health Services_HIP               SAGAMORE_DRG  
4                  Medicaid_United_Medicaid              IUHPS Select_  
...                                     ...                        ...  
21714       Medicaid_Miscellaneous_Medicaid                  Multiplan  
21715                   SIHO_Union Hospital                  Multiplan  
21716               Medicare_Aetna_Medicare                  Multiplan  
21717              Medicaid_Anthem_Medicaid                  Multiplan  
21718                   SIHO_Union Hospital               SAGAMORE_DRG  

[21719 rows x 6 columns]

In [10]:
%%sql 
COPY (
WITH max_min_charges AS (
  SELECT
    cpt_hcpcs_drg AS billing_code, 
    description,
    MIN(negotiated_rate) AS min_charge,
    MAX(negotiated_rate) AS max_charge
  FROM '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet'
  -- WHERE negotiated_rate > 158
  WHERE negotiated_rate > 0
  GROUP BY cpt_hcpcs_drg, description
),

min_max_names AS (
  SELECT
    mmc.billing_code,
    mmc.description, 
    mmc.min_charge,
    mmc.max_charge,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate ASC) AS name_minimum,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate DESC) AS name_maximum
  FROM max_min_charges mmc
  JOIN '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet' sc
  ON mmc.billing_code = sc.cpt_hcpcs_drg AND mmc.description = sc.description
)

SELECT DISTINCT
  billing_code,
  description,
  min_charge,
  max_charge,
  name_minimum,
  name_maximum
FROM min_max_names
) TO '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges_analysis.parquet' (FORMAT 'parquet');

Empty DataFrame
Columns: [Success]
Index: []

In [11]:
%%sql 
SELECT * FROM '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges_analysis.parquet'

billing_code                                       description  \
0            41821                              EXCISION OF GUM FLAP   
1            420-3                                          Diabetes   
2            43770                        LAP PLACE GASTR ADJ DEVICE   
3            447-2  Other Kidney, Urinary Tract & Related Procedures   
4            37607                    LIGATION/BAND ANGIO AV FISTULA   
...            ...                                               ...   
21714        S5552                  INSULIN INTERMED ACTING; 5 UNITS   
21715        0427T  INSJ/RPLC NSTIM SYSTEM SLEEP APNEA PLS GENERATOR   
21716        10008                           FNA BX W/FL GDN EA ADDL   
21717        21485                     CLOSED TX TEMPMAND DISLC COMP   
21718        24301                            MUSCLE/TENDON TRANSFER   

       min_charge  max_charge                               name_minimum  \
0          153.27    11056.50       Medicaid_Managed Health Services_HIP   
1            0.41    13397.17                        SIHO_Union Hospital   
2          153.27    23533.02  Medicaid_Managed Health Services_Medicaid   
3            0.41    21359.38                        SIHO_Union Hospital   
4         2878.56    10850.00               Medicare_Anthem_OON_Medicare   
...           ...         ...                                        ...   
21714        1.80        1.80                    Medicaid_Caresource_HIP   
21715      153.27    76028.46                        Medicaid_Anthem_HIP   
21716     1203.76     7394.00               Medicare_Humana_HMO_Medicare   
21717      295.62     7480.00       Medicaid_Managed Health Services_HIP   
21718      153.27    17129.63               Medicaid_Caresource_Medicaid   

                          name_maximum  
0      Medicaid_Miscellaneous_Medicaid  
1           Medicaid_Self Pay_Medicaid  
2                IUHPS Narrow Network_  
3           Medicaid_Self Pay_Medicaid  
4                          ANTHEM_TRAD  
...                                ...  
21714                       ANTHEM_PPO  
21715            IUHPS Narrow Network_  
21716                  ENCORE_COMBINED  
21717                      ANTHEM_TRAD  
21718            IUHPS Narrow Network_  

[21719 rows x 6 columns]

In [12]:
!ls -lh /Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges_analysis.parquet

-rw-r--r--  1 me  staff   791K Sep  7 17:18 /Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges_analysis.parquet


In [13]:
!cp /Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges_analysis.parquet /Users/me/projects/payless.health/docs/public/data/

# Claude Prompt to make visualization

````
```
meta:
  title: Hospital Charge Data
  description: An interactive dashboard of hospital charge data 

data:
  charges: {file: data/231689692_Lehigh_Valley_Hospital_StandardCharges_analysis.parquet}
  dd: 
    - { u: 0, v: 0 }
    - { u: 500000, v: 500000 }

hconcat:
- vconcat:
  - hconcat:
    - input: menu
      label: Insurance_minimum
      as: $query  
      from: charges
      column: name_minimum
    - input: search 
      label: description
      as: $query
      from: charges
      column: description
      type: contains
  - vspace: 10  
  - plot:
    - mark: dot
      data: {from: charges, filterBy: $query}
      x: min_charge
      y: max_charge
      opacity: 0.1
      fill: name_minimum
      r: 2      
    - mark: regressionY
      data: {from: charges, filterBy: $query}
      x: min_charge  
      y: max_charge
    - select: intervalXY
      as: $query
      brush: { fillOpacity: 0, stroke: black }    
    - mark: lineY
      data: { from: dd }
      x: u
      y: v
      stroke: red
    margins: { left: 60, top: 20, right: 60 }
    xyDomain: Fixed 
    width: 590
    height: 350
    yDomain: [0, 100000]
    xDomain: [0, 100000]
    yGrid: true
    xGrid: true
  - vspace: 5
  - input: table 
    from: charges
    filterBy: $query
    maxWidth: 700
    columns:
      - billing_code
      - description 
      - min_charge
      - max_charge
      - name_minimum
      - name_maximum
    width:  
      billing_code: 60
      description: 200
      min_charge: 90
      max_charge: 90
      name_maximum: 120
      name_minimum: 120
  ```

  Please use the above YAML file and rewrite it for the file `data/351955872_indiana-university-health-inc._standardcharges_analysis.parquet` that has the following column names: `billing_code, description, min_charge, max_charge, name_minimum, name_maximum` 
  ````
